stdlib/public/core/StringObject.swift

```diff
@@ -1219,16 +1282,40 @@ extension _StringObject {
     // Small strings nul-terminate when spilling for contiguous access
     if isSmall { return true }
 
-    // TODO(String performance): Use performance flag, which could be more
-    // inclusive. For now, we only know native strings and small strings (when
-    // accessed) are. We could also know about some shared strings.
-
-    return largeFastIsTailAllocated
+    return largeFastIsTailAllocated || _countAndFlags.isNullTerminated
```

Comment:
I assume `isNullTerminated` will always be set for `largeFastIsTailAllocated`? Can this condition actually just be `_countAndFlags.isNullTerminated`?

Reply:
It isn't currently, since we can derive that information in other ways (it's only ever accessed through this method) and I was trying to not perturb the non-shared String fast paths, but it would be nice to have it consistent. I'll take a look at doing that.

Reply:
Hmm. It's easy enough to do, but looking at it more closely I'm concerned about the implications of deleting the largeFastIsTailAllocated check: this is all inlinable code, but uses of it are not. Old apps will not be setting this flag, so if we rely on it exclusively, we'll be changing their behavior. I don't think we should take that risk. I'll make sure the flag gets set properly for non-immortal Strings going forward though.

Reply:
Just pushed those changes